Recording and saving an audio file causes a crash

#### Software and hardware versions 
f0d5fa2 on sms_submission branch. Android 7.0 on Moto G5 Plus

#### Problem description
Recording and saving an audio file causes a crash

#### Steps to reproduce the problem
* Install the [Sony Audio Recorder](https://play.google.com/store/apps/details?id=com.sonymobile.androidapp.audiorecorder)
* Download the All Widgets form and navigate to the Audio widget
* Record sound (and grant any necessary permissions)
* Save sound and press Done.
* Collect crashes

#### Other information 
```
06-29 13:41:41.508 27437 27437 E AndroidRuntime: java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=3, result=-1, data=Intent { dat=content://com.sonymobile.androidapp.audiorecorder.content/voice recordings/music recordings/Audio recording 2018-06-29 13-41-35.wav typ=audio/raw flg=0x1 }} to activity {org.odk.collect.android/org.odk.collect.android.activities.FormEntryActivity}: java.lang.IllegalArgumentException: column '_data' does not exist
06-29 13:41:41.508 27437 27437 E AndroidRuntime: 	at android.app.ActivityThread.deliverResults(ActivityThread.java:4094)
06-29 13:41:41.508 27437 27437 E AndroidRuntime: 	at android.app.ActivityThread.handleSendResult(ActivityThread.java:4137)
06-29 13:41:41.508 27437 27437 E AndroidRuntime: 	at android.app.ActivityThread.-wrap20(ActivityThread.java)
06-29 13:41:41.508 27437 27437 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1529)
06-29 13:41:41.508 27437 27437 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:102)
06-29 13:41:41.508 27437 27437 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:154)
06-29 13:41:41.508 27437 27437 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:6123)
06-29 13:41:41.508 27437 27437 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
06-29 13:41:41.508 27437 27437 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:867)
06-29 13:41:41.508 27437 27437 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:757)
06-29 13:41:41.508 27437 27437 E AndroidRuntime: Caused by: java.lang.IllegalArgumentException: column '_data' does not exist
06-29 13:41:41.508 27437 27437 E AndroidRuntime: 	at android.database.AbstractCursor.getColumnIndexOrThrow(AbstractCursor.java:333)
06-29 13:41:41.508 27437 27437 E AndroidRuntime: 	at android.database.CursorWrapper.getColumnIndexOrThrow(CursorWrapper.java:87)
06-29 13:41:41.508 27437 27437 E AndroidRuntime: 	at org.odk.collect.android.utilities.MediaUtils.getDataColumn(MediaUtils.java:648)
06-29 13:41:41.508 27437 27437 E AndroidRuntime: 	at org.odk.collect.android.activities.FormEntryActivity.onActivityResult(FormEntryActivity.java:765)
06-29 13:41:41.508 27437 27437 E AndroidRuntime: 	at android.app.Activity.dispatchActivityResult(Activity.java:6931)
06-29 13:41:41.508 27437 27437 E AndroidRuntime: 	at android.app.ActivityThread.deliverResults(ActivityThread.java:4090)
06-29 13:41:41.508 27437 27437 E AndroidRuntime: 	... 9 more
06-29 13:41:41.512  2412  2448 W ActivityManager:   Force finishing activity org.odk.collect.android/.activities.FormEntryActivity
```

Perhaps related to #705, but the stack trace is different.

--- a/collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java
+++ b/collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java
@@ -62,6 +62,7 @@ import com.google.common.collect.ImmutableList;
 import com.google.zxing.integration.android.IntentIntegrator;
 import com.google.zxing.integration.android.IntentResult;
 
+import org.apache.commons.io.IOUtils;
 import org.javarosa.core.model.FormDef;
 import org.javarosa.core.model.FormIndex;
 import org.javarosa.core.model.data.IAnswerData;
@@ -127,6 +128,10 @@ import org.odk.collect.android.widgets.RangeWidget;
 import org.odk.collect.android.widgets.StringWidget;
 
 import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
 import java.text.SimpleDateFormat;
 import java.util.Calendar;
 import java.util.HashMap;
@@ -802,8 +807,37 @@ public class FormEntryActivity extends CollectAbstractActivity implements Animat
 
                 break;
             case RequestCodes.AUDIO_CAPTURE:
-            case RequestCodes.VIDEO_CAPTURE:
+                /*
+                  Probably this approach should be used in all cases to get a file from an uri.
+                  The approach which was used before and which is still used in other places
+                  might be faulty because sometimes _data column might be not provided in an uri.
+                  e.g. https://github.com/opendatakit/collect/issues/705
+                  Let's test it here and then we can use the same code in other places if it works well.
+                 */
                 Uri mediaUri = intent.getData();
+                if (mediaUri != null) {
+                    String filePath =
+                            formController.getInstanceFile().getParent()
+                                    + File.separator
+                                    + System.currentTimeMillis()
+                                    + "."
+                                    + ContentResolverHelper.getFileExtensionFromUri(mediaUri);
+                    try {
+                        InputStream inputStream = getContentResolver().openInputStream(mediaUri);
+                        if (inputStream != null) {
+                            OutputStream outputStream = new FileOutputStream(new File(filePath));
+                            IOUtils.copy(inputStream, outputStream);
+                            inputStream.close();
+                            outputStream.close();
+                            saveFileAnswer(new File(filePath));
+                        }
+                    } catch (IOException e) {
+                        Timber.e(e);
+                    }
+                }
+                break;
+            case RequestCodes.VIDEO_CAPTURE:
+                mediaUri = intent.getData();
                 saveFileAnswer(mediaUri);
                 String filePath = MediaUtils.getDataColumn(this, mediaUri, null, null);
                 if (filePath != null) {
@@ -939,7 +973,7 @@ public class FormEntryActivity extends CollectAbstractActivity implements Animat
         return questionWidget;
     }
 
-    private void saveFileAnswer(Uri media) {
+    private void saveFileAnswer(Object media) {
         // For audio/video capture/chooser, we get the URI from the content
         // provider
         // then the widget copies the file and makes a new entry in the

